modules/docs/docgen/docParser.ts

```diff
@@ -0,0 +1,1777 @@
+}
+
+export const defaultJSDoc: JSDoc = {
+  description: '',
+  tags: {},
+  declarations: [],
+};
+
+export function getFullJsDocComment(checker: ts.TypeChecker, symbol: ts.Symbol) {
+  if (symbol.getDocumentationComment === undefined) {
+    return defaultJSDoc;
+  }
+
+  let mainComment = ts.displayPartsToString(symbol.getDocumentationComment(checker));
+
+  if (mainComment) {
+    mainComment = mainComment.replace(/\r\n/g, '\n');
+  }
+
+  const tags = symbol.getJsDocTags() || [];
+
+  const tagMap: Record<string, string> = {};
+
+  tags.forEach(tag => {
+    const trimmedText = (tag.text || '').trim();
+    const currentValue = tagMap[tag.name];
+    tagMap[tag.name] = currentValue ? currentValue + '\n' + trimmedText : trimmedText;
+  });
+
+  return {
+    description: mainComment,
+    declarations: (symbol?.declarations ?? []).map(d => ({
+      name: symbol?.name || '',
+      filePath: d.getSourceFile().fileName,
+    })),
+    tags: tagMap,
+  };
+}
+
+export function findDocComment(checker: ts.TypeChecker, symbol?: ts.Symbol): JSDoc {
+  if (symbol) {
+    const comment = getFullJsDocComment(checker, symbol);
+    if (comment.description || comment.declarations.length || comment.tags.default) {
+      return comment;
+    }
+
+    const rootSymbols = checker.getRootSymbols(symbol);
+    const commentsOnRootSymbols = rootSymbols
+      .filter(x => x !== symbol)
+      .map(x => getFullJsDocComment(checker, x))
+      .filter(x => !!x.description || !!comment.tags.default);
+
+    if (commentsOnRootSymbols.length) {
+      return commentsOnRootSymbols[0];
+    }
+  }
+
+  return defaultJSDoc;
+}
+
+/**
+ * Attempt to get the name of a declaration or expression
+ */
+export function getNodeName(node: ts.Expression | ts.Declaration): string | undefined {
+  const name = ts.getNameOfDeclaration(node);
+  if (name && 'text' in name) {
+    return name.text;
+  }
+
+  return;
+}
+
+export function filterObjectProperties(value: any): value is ObjectProperty {
+  return value.kind === 'property';
+}
+
+export function filterObjectTypeParameters(value: any): value is TypeParameter {
+  return value.kind === 'typeParameter';
+}
+
+export function filterFunctionParameters(value: any): value is FunctionParameter {
+  return value.kind === 'parameter';
+}
+
+// https://github.com/dsherret/ts-ast-viewer/blob/c71e238123d972bae889b3829e23b44f39d8d5c2/site/src/components/PropertiesViewer.tsx#L172
+export function getSymbolFromNode(checker: ts.TypeChecker, node: ts.Node): ts.Symbol | undefined {
+  return (node as any).symbol || checker.getSymbolAtLocation(node);
+}
+
+export function getPackageName(fileName: string): string {
+  const match = fileName.match(/modules\/([a-z-]+)\//i);
+
+  if (match) {
+    return match[1];
+  }
+
+  return 'react';
+}
+
+export function isObject(type: ts.Type): type is ts.ObjectType {
+  return !!(type.flags & ts.TypeFlags.Object);
```

Comment:
A bitwise AND!? Now I've seen it all.

Reply:
That's how bit flags work. Sometimes Typescript has guard functions and sometimes I have to use bitwise operators.